grammars/silver/compiler/extension/tuple/README.md

```diff
@@ -0,0 +1,43 @@
+
+## Type
+
+We want to properly express error messages on tuples, e.g. rather than printing some "... has actual type `Pair <Integer Pair <String Integer>>`," we would like to notify the user "... has actual type `(Integer, String, Integer)`." In order to do this, we need to create a new `appType` aspect production on the `Type` nonterminal (originally defined in `compiler/definition/type/Type.sv`) that can assist in handling tuples. Then, we also need a new `tupleType` production on the `Type` nonterminal to translate our "tuple type representation" created by `appType` into the appropriate `Pair` type. (Note here that "tuple type representation" is in quotes because it is less of a strict representation and more of a transitional notation for its ultimate representation as a nested `Pair` type).
+
+We define the `tupleType` production to expect a list of `[Type]` that we call `ts` in order to appropriately forward to the `Pair` type construction. Thus, we define a synthesized attribute called `tupleElems` of type `[Type]` that occurs on nonterminal `Type`. (*Allie Thoughts: This seems nonobvious and I would like to expand. Additionally, it seems a bit strange for `tupleElems::[Type]` to occur on `Type`, especially since tuples are the only things that use this. I think this might need some further justification*). 
+
+Since `tupleElems` occurs on `Type`, we can use the `appType` production to append the individual types (like `Integer`) that may be in the the nested pairs into the `tupleElems` list. It is notable that the `appType` production represents the application of a constructor type (like `Pair`), with constructor `c` and argument `a`. In our new production, we pattern match on the `baseType`of `c` and assign the appropriate value to the `tupleElems` of the top-level `Type`. If the `baseType` of the constructor `c` is a `Pair`, then we know its `argTypes` should have only a single element, the `fst` element in the pair. We then append the `tupleElems` of the argument `a` to the list `argTypes` of `c`, which will either recurse on `a` if it is also the application of a constructor type. If it is not, its `tupleElems` will simply be assigned to the singleton list of the top-level as in the `default production`. 
+
+--
+
+Allie has a few thoughts:
+1. If we use `tupleElems` the way that it is currently implemented, won't there be a `tupleElems` assigned for every single application of a constructor type, even if it is not a pair? This does not seem ideal...
+2. I am not sure if I entirely understand the default production. For tuple purposes, I think it is used to assign a value to `tupleElems` in the case when we have the last element type we need to append?
```

Comment:
The default production is shorthand for writing the same equation on every production other than the ones we really care about.  If we have `top.attr = 0;` in the default production, this is the equation in all the productions in the grammar where we *don't* write an equation.

By putting `top.tupleElems = [top];` in the default production, you're saying that the list of types in the "tuple" represented by that type is simply the type itself.